src/blocks/ibus.rs

```diff
@@ -0,0 +1,185 @@
+use std::env;
+use std::fs::File;
+use std::io::prelude::*;
+use std::sync::{Arc, Mutex};
+use std::thread;
+use std::time::{Duration, Instant};
+use chan::Sender;
+
+use block::{Block, ConfigBlock};
+use config::Config;
+use errors::*;
+use input::I3BarEvent;
+use scheduler::Task;
+use uuid::Uuid;
+use widgets::text::TextWidget;
+use widget::I3BarWidget;
+
+extern crate dbus;
+extern crate regex;
+use self::dbus::{Connection, ConnectionItem, arg};
+use self::regex::Regex;
+
+pub struct IBus {
+    id: String,
+    text: TextWidget,
+    engine: Arc<Mutex<String>>,
+}
+
+#[derive(Deserialize, Debug, Default, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct IBusConfig {
+    // TODO: Implement this.
+    // Set to display engine name as the two letter country abbreviation, e.g. "jp".
+    #[serde(default = "IBusConfig::default_abbreviate")]
+    pub as_icon: bool,
+}
+
+impl IBusConfig {
+    fn default_abbreviate() -> bool {
+        true
+    }
+}
+
+impl ConfigBlock for IBus {
+    type Config = IBusConfig;
+
+    fn new(_block_config: Self::Config, config: Config, send: Sender<Task>) -> Result<Self> {
+        let id: String = Uuid::new_v4().simple().to_string();
+        let id_copy = id.clone();
+
+        let ibus_address = get_ibus_address()
+            .block_error("ibus", "Could not get ibus address")?;
+        let c = Connection::open_private(&ibus_address)
+            .block_error("ibus", &format!("Failed to establish D-Bus connection to {}", ibus_address))?;
+        let p = c.with_path("org.freedesktop.IBus", "/org/freedesktop/IBus", 5000);
+        use blocks::dbus::stdintf::org_freedesktop_dbus::Properties;
+        let info: arg::Variant<Box<arg::RefArg>> = p.get("org.freedesktop.IBus", "GlobalEngine")
+            .block_error("ibus", "Couldn't get ibus address")?;
+
+        // `info` should contain something containing an array with the contents as such:
+        // [name, longname, description, language, license, author, icon, layout, layout_variant, layout_option, rank, hotkeys, symbol, setup, version, textdomain, icon_prop_key]
+        // Refer to: https://github.com/ibus/ibus/blob/7cef5bf572596361bc502e8fa917569676a80372/src/ibusenginedesc.c
+        // e.g.                   name           longname        description     language
+        // ["IBusEngineDesc", {}, "xkb:us::eng", "English (US)", "English (US)", "en", "GPL", "Peng Huang <shawn.p.huang@gmail.com>", "ibus-keyboard", "us", 99, "", "", "", "", "", "", "", ""]
+        //                         ↑ We will use this element (name) as it is what GlobalEngineChanged signal returns.
+        let current = info.0.as_iter().unwrap().nth(2).unwrap().as_str().unwrap_or("??");
```

Comment:
Can you avoid using `unwrap()` here (and really anywhere) in favour of `block_error()` and friends?

Reply:
Have done this now. Only these ones are left:

```rust
^^/i3status-rust >>> rg "unwrap()" src/blocks/ibus.rs
// Do these need to be replaced? All the other blocks using D-Bus were like this:
77:            let c = Connection::open_private(&ibus_address).unwrap();
80:            ).unwrap();
// this should be OK
// "If another user of this mutex panicked while holding the mutex, then this call will return an error once the mutex is acquired."
84:                            let mut engine = engine_original.lock().unwrap();
// These return Options on the received dbus message item. Don't see why these would error..
134:    if &*m.interface().unwrap() != "org.freedesktop.IBus" { return None };
135:    if &*m.member().unwrap() != "GlobalEngineChanged" { return None };

// These will only fail if the given regex expression is invalid.
176:    let re = Regex::new(r"^:(\d{1})$").unwrap();
189:    let re = Regex::new(r"IBUS_ADDRESS=(.*),guid").unwrap();

Reply:
Done. 

By the way, off-topic to this PR but there's about 50 or so `unwrap()`s used in the other blocks in the codebase. Ideally should these be looked at and replaced one day? If some still remain after refactoring then do we assume nothing is obvious and add a comment to each one explaining why they will not fail?